certora/specs/RewardsDistributor.spec

```diff
@@ -0,0 +1,85 @@
+    function claimed(address) external returns uint256 envfree;
+    function claim(address, uint256, bytes32[]) external envfree;
+
+    function MerkleTrees.getValue(address, address) external returns uint256 envfree;
+    function MerkleTrees.getHash(address, address) external returns bytes32 envfree;
+    function MerkleTrees.wellFormedPath(address, address, bytes32[]) external envfree;
+
+    function MorphoToken.balanceOf(address) external returns uint256 envfree;
+}
+
+// Check an account claimed amount is correctly updated.
+rule updatedClaimedAmount(address _account, uint256 _claimable, bytes32[] _proof) {
+    claim(_account, _claimable, _proof);
+
+    assert _claimable == claimed(_account);
+}
+
+// Check an account can only claim greater amounts each time.
+rule increasingClaimedAmounts(address _account, uint256 _claimable, bytes32[] _proof) {
+    uint256 claimed = claimed(_account);
+
+    claim(_account, _claimable, _proof);
+
+    assert _claimable > claimed;
+}
+
+// Check that claiming twice is equivalent to claiming once with the last amount.
+rule claimTwice(address _account, uint256 _claim1, uint256 _claim2) {
+    storage initStorage = lastStorage;
+
+    bytes32[] _proof1; bytes32[] _proof2;
+    claim(_account, _claim1, _proof1);
+    claim(_account, _claim2, _proof2);
+
+    storage afterBothStorage = lastStorage;
+
+    bytes32[] _proof;
+    claim(_account, _claim2, _proof) at initStorage;
+
+    assert lastStorage == afterBothStorage;
+}
+
+// Check that the transferred amount is equal to the claimed amount minus the previous claimed amount.
+rule transferredTokens(address _account, uint256 _claimable, bytes32[] _proof) {
+    // Assume that the rewards distributor itself is not receiving the tokens, to simplify this rule.
+    require _account != currentContract;
+
+    uint256 balanceBefore = MorphoToken.balanceOf(_account);
+    uint256 claimedBefore = claimed(_account);
+
+    // Safe require because the sum is capped by the total supply.
+    require balanceBefore + MorphoToken.balanceOf(currentContract) < 2^256;
+
+    claim(_account, _claimable, _proof);
+
+    uint256 balanceAfter = MorphoToken.balanceOf(_account);
+
+    assert balanceAfter - balanceBefore == _claimable - claimedBefore;
+}
+
+rule claimCorrectness(address _account, uint256 _claimable, bytes32[] _proof) {
+    address prevTree; address prevNode;
+    address currTree; address currNode;
+
+    // Assume that prevRoot and currRoot correspond to prevTree and currTree.
+    require MerkleTrees.getHash(prevTree, prevNode) == prevRoot();
+    require MerkleTrees.getHash(currTree, currNode) == currRoot();
+
+    // No need to make sure that currNode (resp prevNode) is equal to currRoot (resp prevRoot): one can pass an internal node instead.
+
+    // Assume that prevTree and currTree are well-formed.
+    MerkleTrees.wellFormedPath(prevTree, prevNode, _proof);
+    MerkleTrees.wellFormedPath(currTree, currNode, _proof);
```

Comment:
In CVL when you call a function, it assumes that it does not revert by default. You can change that default with a trailing `@withrevert`. Here we use that default behavior, and that function would only revert because of its `require`. But it's not actually making the verification less general because the require is a verfied invariant, so we know we only remove false counter-examples